Initial Task Executor implementation Task(on:), addTask(on:) etc.

ktoso opened the pull request:
Pending proposal and Swift Evolution process, do not merge yet.

This generally will allow tasks to be launched/spawned immediately on a specific executor or actor, and handle all the associated isolation.

Resolves rdar://107072140

ktoso commented on test/Concurrency/Runtime/async_custom_task_executor.swift:
```diff
@@ -0,0 +1,99 @@
+        // do nothing
+      }
+      assert(MyGlobalActor.shared.executor.enqueueCount == 1)
+    }
```
^^^ Current list of APIs; the global actor ones we'll have to figure out better and it'd be nice to avoid the `actor in` dance for actors.

ktoso committed:
emitting hop

ktoso committed:
nonisolated funcs run on preferred executor

ktoso committed:
show hopping in async, nonisolated, closure

ktoso committed:
basic withExecutor implemented

ktoso committed:
dont change function prologue

ktoso committed:
prepare for handling Executor and not just SerialExecutor

ktoso committed:
task excutor pref inheritance for task group

ktoso committed:
Task Executors: Prepare for new TaskExecutor protocol & builtins

Rename ExecutorRef to SerialExecutorRef

Apply suggestions from code review

Co-authored-by: John McCall <rjmccall@gmail.com>

remove legacy ExecutorRef typealias

remove buildOrdinaryTaskExecutorRef until we need it

ktoso committed:
adjust to new SerialExecutor and TaskExecutor split

ktoso committed:
nonisolated func worked but on serial executor

ktoso committed:
first time passing async_task_executor_default_actor_funcs.swift

ktoso committed:
fixed async_task_executor_nonisolated_async_fun

ktoso committed:
cleaning up Concurrency/Runtime/async_task_executor_withExecutor.swift

ktoso committed:
implement disabling task executor preference

ktoso committed:
complete structured concurrency

ktoso committed:
implemented UnsafeCurrentTask.unownedTaskExecutor

ktoso committed:
correct switching logic

ktoso commented:
@swift-ci please build toolchain

ktoso committed:
Cleanup tests for UnsafeCurrentTask

ktoso commented:
@swift-ci please build toolchain macOS

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please build toolchain macOS

ktoso committed:
remove incorrect test

ktoso committed:
Cleanup and document

ktoso committed:
add missing APIs for TaskGroup and cleanup

ktoso committed:
correct logic for TaskGroup addTask(on: nil)

ktoso commented:
@swift-ci please smoke test

rjmccall commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -2485,6 +2486,9 @@ class JobFlags : public FlagSet<uint32_t> {
                                 task_setIsAsyncLetTask)
+  FLAGSET_DEFINE_FLAG_ACCESSORS(Task_HasInheritedExecutorPreference,
+                                task_hasInitialTaskExecutorPreferenceRecord,
+                                task_setHasInitialTaskExecutorPreferenceRecord)
```
Two different names for the same concept

ktoso replied:
So the "initial" is slightly different. It is one we create from the "on:" parameters and we need to know if we did so in order to destroy it properly as we destroy the task in "complete". Other preference records just are destroyed using an adequate "pop" pair to the "push" calls.

ktoso replied:
Oh whoops i see Inherited/Initial, yes it should be Initial -- thank you

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -410,6 +415,19 @@ class AsyncTask : public Job {
+  void pushInitialTaskExecutorPreference(TaskExecutorRef preferred);
+
+  /// WARNING: Only to be used during task completion (destroy).
+  /// This is because this API assumes that we have an executor prefer
```
Unfinished comm

ktoso replied:
Fixed sorry about that

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -513,6 +531,19 @@ class AsyncTask : public Job {
+  /// setting it during creation (`Task(on:...)`).
+  ///
+  /// This means that during task tear down the record should be deallocated
+  /// because it was not set with a
```
Another unfinished comme

ktoso replied:
Thanks, fixed in 098ff1880ae0a3e1c713084e482919c1c63c8f01 and looked at all other comments as well.

rjmccall commented on include/swift/ABI/TaskOptions.h:
```diff
@@ -78,23 +78,25 @@ class TaskGroupTaskOptionRecord : public TaskOptionRecord {
+/// or the default executor should be used.
+///
+/// Lack of this option usually means that the global concurrent executor, or the executor of the enclosing actor
+/// will be used.
```
Please word-wrap comments at ~80 columns

ktoso replied:
Ok I'll make sure to clang format it all

ktoso replied:
Okey, yeah I wasn't sure if we want to hardcode this assumption here -- you are right that it is guaranteed currently to be a `{0, 0}` basically (SerialExecutorRef), so okey if we want to assume this I removed the function and replaced it with:

```
@available(SwiftStdlib 9999, *)
@usableFromInline
internal func _getGenericSerialExecutor() -> Builtin.Executor {
  // The `SerialExecutorRef` to "default generic executor" is guaranteed
  // to be a zero value;
  //
  // As the runtime relies on this in multiple places,
  // so instead of a runtime call to get this executor ref, we bitcast a "zero"
  // of expected size to the builtin executor type.
  unsafeBitCast((UInt(0), UInt(0)), to: Builtin.Executor.self)
}
```

Is that the right UInt to use here...? In C++ it is:

```
  HeapObject *Identity; // Not necessarily Swift reference-countable
  uintptr_t Implementation;
```

rjmccall commented on include/swift/ABI/TaskStatus.h:
```diff
@@ -276,6 +276,25 @@ class EscalationNotificationStatusRecord : public TaskStatusRecord {
 
+/// This record signifies that the task has an executor preference.
+/// This preference may be added or removed at runtime, e.g. when multiple `withTaskExecutor { ... }`
+/// blocks are nested, they add more executor preferences.
```
Suggestion: Any number of these preferences may be present at runtime, and the
innermost preference takes priority.

ktoso replied:
Good note, added it 👍

rjmccall commented on include/swift/AST/Builtins.def:
```diff
@@ -983,6 +986,34 @@ BUILTIN_MISC_OPERATION_WITH_SILGEN(CreateAsyncTask,
+/// Build a Builtin.Executor value from an "ordinary" task executor
+/// reference.
+BUILTIN_MISC_OPERATION_WITH_SILGEN(BuildOrdinaryTaskExecutorRef,
+    "buildOrdinaryTaskExecutorRef", "n", Special)
```
Did you end up needing these builtins for some reason?

ktoso replied:
Hmmm, I needed them while I was emitting "get the task executor" in SIL in nonisolated functions but that we moved away from. I think the current approach is the right one (wanted to confirm with this review request) so let me remove all the not used things in a commit:

ktoso replied:
Github seems to have lost my comment -- yes, we need these because: 

```
struct UnownedTaskExecutor {

  @inlinable
  public init<E: TaskExecutor>(ordinary executor: __shared E) {
    #if $BuiltinBuildTaskExecutor
    self.executor = Builtin.buildOrdinaryTaskExecutorRef(executor)
    #else
    fatalError("Swift compiler is incompatible with this SDK version")
    #endif
  }
}
```

We arguably could "wing it" and use the `buildOrdinaryExecutorRef` though I'm unsure about the proliferation of pretending as if they're the same... Internally the representation is the same though, that is true -- the memory layout of SerialExecutorRef and TaskExecutorRef is the same, at least currently.

rjmccall commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -900,6 +918,20 @@ SerialExecutorRef swift_task_getCurrentExecutor(void);
 
+/// Return the generic (default global concurrent) executor reference.
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+SerialExecutorRef swift_task_getGenericExecutor(void);
```
If you're adding these new functions to the runtime ABI, please note which Swift release added them.

But this function in particular I think we don't need — there's an ABI-guaranteed way of constructing this already (IIRC it's the all-zeroes value).

ktoso replied:
Yes, it's `0, 0` so I think I could bitcast cast... I'll give that a go so we can avoid runtime functions here.
I wasn't sure if we want to bake in this assumption but I guess it already is baked in 🤔

ktoso replied:
I did the same for `_getUndefinedTaskExecutor` as well, to avoid runtime calls: https://github.com/apple/swift/pull/68793/commits/1318d3ead1526e596dd302e434d43622d8aa45c0

rjmccall commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -900,6 +918,20 @@ SerialExecutorRef swift_task_getCurrentExecutor(void);
+/// It can be used used to store a record indicating that
+/// there is no task executor preference in the current scope.
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+TaskExecutorRef swift_task_getUndefinedTaskExecutor(void);
```
Similarly, I don't think we need a function for this, because we really want this to be guaranteed to be the all-zeroes value.

ktoso replied:
Okey gotcha so I'll just bitcast zeroes and avoid the runtime call. Thanks!

ktoso replied:
Done, and removed the runtime func -- explained more here: https://github.com/apple/swift/pull/68793#discussion_r1398654044

rjmccall commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -900,6 +918,20 @@ SerialExecutorRef swift_task_getCurrentExecutor(void);
+TaskExecutorRef swift_task_getUndefinedTaskExecutor(void);
+
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+TaskExecutorRef swift_task_getPreferredTaskExecutor(void);
```
This really needs documentation explaining the constraints on how it can be used.

ktoso replied:
Added in: https://github.com/apple/swift/pull/68793/commits/10cadfa2a876bbe3d9a3ae736b186dc6425cc08a I think that covers the "be careful" side of things. Anything more we should add?

rjmccall reviewed (changes requested):
Leaving off there for now

ktoso committed:
Address review comments

ktoso committed:
clang-format

ktoso committed:
clang-format the task executor work

ktoso committed:
properly guard the BuiltinBuildTaskExecutor

ktoso committed:
Remove runtime func to get generic executor; cast instead

ktoso committed:
undo changes to
test/Concurrency/Runtime/async_task_withUnsafeCurrentTask.swift

ktoso committed:
fix comment

ktoso committed:
Include comment about multiple instances of preference record

ktoso committed:
remove the swift_task_getUndefinedTaskExecutor runtime call; impl with cast

ktoso committed:
document swift_task_getPreferredTaskExecutor

ktoso committed:
more documentation of new Runtime APIs

ktoso commented:
(for after thanks-giving):

I noticed the following warning and unsure what to do about it

```
/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/GlobalExecutor.cpp:139:19: warning: '_task_serialExecutor_getExecutorRef' has C-linkage specified, but returns user-defined type 'SerialExecutorRef' which is incompatible with C [-Wreturn-type-c-linkage]
  139 | SerialExecutorRef _task_serialExecutor_getExecutorRef(
      |                   ^
/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/GlobalExecutor.cpp:147:17: warning: '_task_executor_getTaskExecutorRef' has C-linkage specified, but returns user-defined type 'TaskExecutorRef' which is incompatible with C [-Wreturn-type-c-linkage]
  147 | TaskExecutorRef _task_executor_getTaskExecutorRef(
      |                 ^
2 warnings generated.
```

ktoso committed:
fix various warnings in tests

ktoso committed:
Avoid adding un-necessary runtime and builtin funcs - HopToExecutor

These were added in earlier commits of the initial task executor work,
however they are not necessary after all -- we get the right hop
behavior just from the task executor stored and picked up in runtime
functions already.

ktoso committed:
allow an executor to be BOTH Serial and Task

ktoso committed:
correct sil test

ktoso commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -541,6 +541,29 @@ bool swift_taskGroup_hasTaskGroupRecord();
+/// signals a bug in record handling by the concurrency library -- a record push
+/// must always be paired with a record pop.
+///
+/// Runtime availability: Swift 9999.
```
We've not been doing comments like this much here tbh, but I think that's what you meant John?

ktoso committed:
docs: runtime availability

ktoso committed:
cleanup unused code, and undo formatting changes

ktoso committed:
Optimize: status flag to avoid scanning records for preference

ktoso commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -525,6 +596,16 @@ static bool swift_task_hasTaskGroupStatusRecordImpl() {
+  //
+  // This is an optimization in order to make the enqueue/run
+  // path of a task avoid excessive work if a task had many records.
+  if (!hasTaskExecutorPreferenceRecord()) {
```
This commit https://github.com/apple/swift/pull/68793/commits/4f16bdc8152e97eca12012278ab8f6882d486ba7 was the last thing I wanted to do here.

Since we query the task executor in functions like `defaultActorDrain` it would be a waste to for every task keep scanning all the records if most of them will not have a preference record.

We use 1 bit flag in the task status to indicate that there will be at least one preference record so we can avoid scanning the records all the time.

If this is a bad tradeoff (I thought making the "normal, no preferences" path be fast is preferable) I can revert this single commit. This is a piece I'd like some feedback on, thank you!

ktoso commented:
@swift-ci please smoke test

ktoso committed:
add missing dispatch import flag

ktoso committed:
fix syntax in compatibility override test

ktoso commented:
@swift-ci please smoke test

ktoso committed:
add task executor test with distributed actor func

ktoso committed:
stability-concurrency-abi required explicit availability on accessor

ktoso committed:
embedded cannot support task executors, make unavailable

ktoso committed:
remove warnings from async_task_executor_structured_concurrency.swift

ktoso commented:
Ok linux all good but missed some embedded tests, fixed now.

ktoso commented:
@swift-ci please smoke test

ktoso committed:
fix test in CompatibilityOverrideConcurrency

ktoso committed:
compatibility overrides for the new task executor push/pop/get methods

ktoso commented:
@swift-ci please smoke test

ktoso committed:
Remove not used RuntimeFunc definition

ktoso commented:
@swift-ci please smoke test

ktoso commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -530,7 +530,7 @@ void swift_asyncLet_consume_throwing(SWIFT_ASYNC_CONTEXT AsyncContext *,
 /// \endcode
 SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
-bool swift_taskGroup_hasTaskGroupRecord();
+bool swift_taskGroup_hasTaskGroupRecord(); // FIXME: not used? we have swift_task_hasTaskGroupStatusRecord
```
I'm wondering if we could remove this 🤔

ktoso commented on stdlib/public/Concurrency/Task+TaskExecutor.swift:
```diff
@@ -0,0 +1,212 @@
+  // the operation, its enqueue will respect the attached executor preference.
+
+  return try await operation()
+}
```
The "temporarily select a task excutor" API

ktoso commented on stdlib/public/Concurrency/Task+TaskExecutor.swift:
```diff
@@ -0,0 +1,212 @@
+    // Create the asynchronous task.
+    let taskExecutorRef = executor.asUnownedTaskExecutor()
+    let (task, _) = Builtin.createAsyncTaskWithExecutor(
+      flags, taskExecutorRef.executor, operation)
```
These `Task(on:)` are the reason for the `Builtin.createAsyncTaskWithExecutor` builtin. And the same is the case for the TaskGroup.addTask APIs

ktoso committed:
No assertion API for TaskExecutor

ktoso commented:
@swift-ci please build toolchain

ktoso commented:
Status update: Ok this is complete from my perspective but since ABI and runtime impacts will want to give John more time to review before we land this. If we need to change things around based on feedback we will before landing 👍

CodaFi commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -100,9 +100,15 @@ public protocol SerialExecutor: Executor {
+/// to execute any work which does not have strict executor requirements.
+///
+/// By setting a task executor preference, either with a ``withTaskExecutor(_:operation:)``, creating
+/// a task with a preference (`Task(on:)`), or
```
Nit: incomplete thought

ktoso replied:
Ouch, thanks for noticing, I hoped I got them all -- seems not. Fixing 👍

ktoso committed:
Underscore _TaskExecutor, addTask, and related APIs

ktoso committed:
Comment formatting and add cut off sentence

ktoso committed:
add Task.detached(_on:) which was missing

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test macOS
